From bae4b9bbdc649d88990857d92e9e6c43c4ec2f3e Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Thu, 15 Sep 2016 21:17:07 +0100
Subject: [PATCH 08/10] applets/status: Correctly port to the upower 0.99 API

This is a serious oversight which means the remove signal would always result
in a segfault, as we're trying to treat a string as a gobject. Sorta figures
it would go bang.

This reenforces the case for using C even further by not masking issues under
a vapi file. This resolves issue #594.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 panel/applets/status/PowerIndicator.vala | 15 ++++++++-------
 vapi/upower-glib-1.0.vapi                |  3 +--
 2 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/panel/applets/status/PowerIndicator.vala b/panel/applets/status/PowerIndicator.vala
index 76fa174..386dd28 100644
--- a/panel/applets/status/PowerIndicator.vala
+++ b/panel/applets/status/PowerIndicator.vala
@@ -149,16 +149,17 @@ public class PowerIndicator : Gtk.Bin
      */
     void on_device_added(Up.Device device)
     {
-        if (devices.contains(device.serial)) {
+        string object_path = device.get_object_path();
+        if (devices.contains(object_path)) {
             /* Treated as a change event */
-            devices.lookup(device.serial).update_ui(device);
+            devices.lookup(object_path).update_ui(device);
             return;
         }
         if (!this.is_interesting(device)) {
             return;
         }
         var icon = new BatteryIcon(device);
-        devices.insert(device.serial, icon);
+        devices.insert(object_path, icon);
         widget.pack_start(icon);
         toggle_show();
     }
@@ -176,13 +177,13 @@ public class PowerIndicator : Gtk.Bin
     /**
      * Remove a device from our display
      */
-    void on_device_removed(Up.Device device) {
-        if (!devices.contains(device.serial)) {
+    void on_device_removed(string object_path) {
+        if (!devices.contains(object_path)) {
             return;
         }
-        unowned BatteryIcon? icon = devices.lookup(device.serial);
+        unowned BatteryIcon? icon = devices.lookup(object_path);
         widget.remove(icon);
-        devices.remove(device.serial);
+        devices.remove(object_path);
         toggle_show();
     }
 
diff --git a/vapi/upower-glib-1.0.vapi b/vapi/upower-glib-1.0.vapi
index 218630f..531c76f 100644
--- a/vapi/upower-glib-1.0.vapi
+++ b/vapi/upower-glib-1.0.vapi
@@ -56,8 +56,7 @@ namespace Up {
 		public bool on_low_battery { get; }
 		public virtual signal void changed ();
 		public virtual signal void device_added (Up.Device device);
-		public virtual signal void device_changed (Up.Device device);
-		public virtual signal void device_removed (Up.Device device);
+		public virtual signal void device_removed (string object_path);
 		public virtual signal void notify_resume (uint sleep_kind);
 		public virtual signal void notify_sleep (uint sleep_kind);
 	}
-- 
2.10.0

